### PR TITLE
feat(DX-374): PinCode Trait support

### PR DESCRIPTION
**Pull request by ses110**
Support PinCode Trait:

We're mapping the query for a PinCodeTrait:
```
PinCodeCredentialDeviceTrait {
      properties {
        maxNumberOfPinCodeCredentials
        supportedPinCodeCredentialNameRange {
          min
          max
        }
        supportedPinCodeRange {
          min
          max
        }

      }
      state {
        pinCodeCredentials {
          reported {
            # there can be large numbers of pin codes, so this also follows the relay spec:
            # https://relay.dev/graphql/connections.htm
            value {
              edges {
                node {
                  name
                  pinCode
                  schedules
                }
              }
            }
            createdAt
            sampledAt
          }
          desired {
            value {
              edges {
                node {
                  name
                  pinCode
                  schedules
                }
              }
            }
            delta {
              edges {
                node {
                  name
                  pinCode
                  schedules
                }
              }
            }
            updatedAt
          }
        }
        pinCodeCredentialList {
          reported {
            sampledAt
            createdAt
            value {
              edges {
                node {
                  name
                  pinCode
                  alwaysActive
                  recurringSchedules {
                    duration
                    rrule
                  }
                  temporarySchedules {
                    startDatetime
                    endDatetime
                  }
                }
              }
            }
          }
          desired {
            updatedAt
            value {
              edges {
                node {
                  name
                  pinCode
                  alwaysActive
                  recurringSchedules {
                    duration
                    rrule
                  }
                  temporarySchedules {
                    startDatetime
                    endDatetime
                  }
                }
              }
            }
          }
        }
      }
    }
```

To this interface in Dart:
```
    expect(pinCodeTrait.pinCodes?[0].name, equals("Admin"));
    expect(pinCodeTrait.pinCodes?[0].pinCode, equals("0070"));
    expect(pinCodeTrait.pinCodes?[0].toString(),
        equals("PinCodeCredential(name: Admin, pinCode: 0070)"));

    expect(pinCodeTrait.pinCodes?[1].name, equals("Plant Lady"));
    expect(pinCodeTrait.pinCodes?[1].pinCode, equals("5161"));

    expect(pinCodeTrait.maxNumberOfCredentials, 10);

    expect(pinCodeTrait.nameLengthRange?.min, 1);
    expect(pinCodeTrait.nameLengthRange?.max, 30);

    expect(pinCodeTrait.pinCodeLengthRange?.min, 1);
    expect(pinCodeTrait.pinCodeLengthRange?.max, 100);
```

### CODE REVIEW

**Comment by ses110** [comment]
Code coverage for file `pin_code_repository.dart` should be near 100% now, but codecov is not reporting coverage for this line:
`PinCodeCredential operator [](int index) => value[index];`
when it is used in `devices_repository_test.dart` on lines 603-608.

File: lib/src/repository/traits/pin_code_repository.dart [hunk 1]
```diff
@@ -0,0 +1,151 @@
...
+}
+
+///
+/// State-related classes representing Pin Code credentials for a PinCode Trait
+///
+class PinCodeCredentials extends State<List<PinCodeCredential>> {
+  PinCodeCredentials(rawValues)
+      : super((PinCodeCredentials).toString(), _toMappedList(rawValues)) {}
+
+  PinCodeCredential operator [](int index) => value[index];
```
Comment (grizzly-atoms) [inline lib/src/repository/traits/pin_code_repository.dart:63]: What would it take to get test coverage here?

Comment (ses110) [reply lib/src/repository/traits/pin_code_repository.dart:63]: I mentioned in my comment here that CodeCov is not reading this line as covered when it should:
https://github.com/Yonomi/yonomi-sdk-dart/pull/81#issuecomment-1231852634

File: test/unit/devices_repository_test.dart [hunk 2]
```diff
@@ -158,17 +170,24 @@ void main() {
               }
             }
           },
+          {
+            "__typename": "NewUnknownTrait",
```
Comment (rlperez) [inline test/unit/devices_repository_test.dart:174]: Any background on this?

File: lib/src/repository/traits/pin_code_repository.dart (hunk 1 above)
Comment (rlperez) [inline lib/src/repository/traits/pin_code_repository.dart:63]: There should be an annotation for ignoring this. You are using it so it should be good...

```
// coverage:ignore-line to ignore one line.
// coverage:ignore-start and // coverage:ignore-end to ignore range of lines inclusive.
// coverage:ignore-file to ignore the whole file.
```

Maybe this? https://stackoverflow.com/a/66171720/674989

File: lib/src/queries/pin_code_credential/make_create_pin_code_credential_request/query.graphql [hunk 3]
```diff
@@ -0,0 +1,28 @@
...
+            pinCodeCredentials {
+              reported { value { edges { node { pinCode }}} sampledAt createdAt }
+              desired { value { edges { node { pinCode }}} updatedAt }
+            }
+          }
+        }
+      }
+    }
+  }
+}
```
Comment (rlperez) [inline lib/src/queries/pin_code_credential/make_create_pin_code_credential_request/query.graphql:28]: Might as well add the new line if you make any other changes

File: test/unit/devices_repository_test.dart (hunk 2 above)
Comment (ses110) [reply test/unit/devices_repository_test.dart:174]: It's to make sure our code can parse any currently-unsupported Trait as the UnknownTrait class

File: lib/src/repository/traits/pin_code_repository.dart (hunk 1 above)
Comment (ses110) [reply lib/src/repository/traits/pin_code_repository.dart:64]: Yeah, CodeCov is not picking up that this is an overridden operator and it's technically being used in the tests.

**Review by grizzly-atoms — APPROVED** [review]

**Pull request merged**